homeassistant/loader.py

```diff
@@ -63,6 +63,69 @@ def manifest_from_legacy_module(domain: str, module: ModuleType) -> Dict:
     }
 
 
+async def _get_custom_components(hass: 'HomeAssistant') -> List['Integration']:
+    """Return list of custom integrations."""
+    try:
+        import custom_components
+    except ImportError:
+        return []
+
+    def get_custom_dirs(paths: List) -> List:
+        return [
+            entry
+            for path in paths
+            for entry in pathlib.Path(path).iterdir()
+            if entry.is_dir()
+        ]
+
+    dirs = await hass.async_add_executor_job(
+        get_custom_dirs, custom_components.__path__)
+
+    components = await asyncio.gather(*[
+        hass.async_add_executor_job(
+            Integration.resolve_from_root,
+            hass,
+            custom_components,
+            comp)
+        for comp in dirs
+    ])
+
+    return [
+        component
+        for component in components
+        if component is not None
+    ]
+
+
+async def get_custom_components(hass: 'HomeAssistant') -> List['Integration']:
+    """Return cached list of custom integrations."""
+    cache = hass.data.get(DATA_CUSTOM_COMPONENTS)
+    if cache is None:
+        cache = hass.data[DATA_CUSTOM_COMPONENTS] = {}
+        cache['lock'] = asyncio.Lock()
+        cache['data'] = None
+    async with cache['lock']:
+        if cache['data'] is None:
+            cache['data'] = await _get_custom_components(hass)
+        return cast(List['Integration'], cache['data'])
```

Comment:
This caching logic could probably be generalized into a decorator. might simplify code elsewhere.

Reply:
To remember how, not convinced it was a good idea.
```
def cached(key):
    def decorator_cached(func):
        @ft.wraps(func)
        async def wrapper_cached(*args, **kwargs):
            hass = args[0]
            cache = hass.data.get(key)
            if cache is None:
                cache = hass.data[key] = {}
                cache['lock'] = asyncio.Lock()
                cache['data'] = None
            async with cache['lock']:
                if cache['data'] is None:
                    cache['data'] = await func(*args, **kwargs)
                return cache['data']
        return wrapper_cached
    return decorator_cached
```